backend/middleware/timeouts/timeouts.go

```diff
@@ -57,18 +59,64 @@ func (m *mid) getDuration(service, method string) time.Duration {
 	return m.defaultTimeout
 }
 
+type unaryHandlerReturn struct {
+	resp interface{}
+	err  error
+}
+
 func (m *mid) UnaryInterceptor() grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 		service, method, ok := middleware.SplitFullMethod(info.FullMethod)
 		if !ok {
 			m.logger.Warn("could not parse gRPC method", zap.String("fullMethod", info.FullMethod))
 		}
 
+		// Create a return channel for the goroutine.
+		resultChan := make(chan unaryHandlerReturn)
+		defer close(resultChan)
+
+		// Create a channel to track when the timeout error has already been returned and the return channel is closed.
+		done := make(chan struct{})
+		defer close(done)
```

Comment:
I understand the intent, but I'm not grokking the implementation. Basically we wait for the handler to respond, even if the timeout has exceeded so we can produce a helpful log message understanding why with a hopeful error message.

Two questions:
1. How is this channel closed after the handler is complete?
2. The handler will get a context with a timeout, it might not respect it and in this case we still wait for the handler to complete. Can we leak go routines here?

Reply:
1. yes, if the channel is closed the handler goroutine will receive https://golang.org/ref/spec#Close
1. ultimately go does not give you any control over the goroutine. there's no way to interrupt or stop it. so yes it would leak if the function never returns https://tip.golang.org/doc/faq#no_goroutine_id. note: this leak is not new to this implementation since the same leak would be present before in the goroutine running the handler https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#servers

Reply:
Nice thanks for the explanation